### PR TITLE
⭐️ Entity embedder interface is here

### PR DESCRIPTION
**Pull request by EssamWisam**
Add an interface for `EntityEmbedder` that can wrap any basic deep learning model in an unsupervised model.

```julia
using MLJFlux
using MLJ
using CategoricalArrays

N = 200
X = (;
    Column1 = repeat(Float32[1.0, 2.0, 3.0, 4.0, 5.0], Int(N / 5)),
    Column2 = categorical(repeat(['a', 'b', 'c', 'd', 'e'], Int(N / 5))),
    Column3 = categorical(repeat(["b", "c", "d", "f", "f"], Int(N / 5)), ordered = true),
    Column4 = repeat(Float32[1.0, 2.0, 3.0, 4.0, 5.0], Int(N / 5)),
    Column5 = randn(Float32, N),
    Column6 = categorical(
        repeat(["group1", "group1", "group2", "group2", "group3"], Int(N / 5)),
    ),
)

y = categorical([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])           # Classification

NeuralNetworkClassifier = @load NeuralNetworkClassifier pkg=MLJFlux

clf = NeuralNetworkClassifier(embedding_dims=Dict(:Column2 => 2, :Column3 => 2))

emb = EntityEmbedder(clf)

mach = machine(emb, X, y)

fit!(mach)

Xnew = transform(mach, X)
Xnew
``` 

### PR Checklist

- [x] Tests are added
- [x] Documentation, if applicable

### CODE REVIEW

**Comment by ablaom** [comment]
Thanks @EssamWisam  for this valuable contribution.

Do we know why buildkite tests are failing?

**Comment by EssamWisam** [comment]
Everything runs fine locally. For buildkite logs:
image

Potential change that triggered this is adding the `rng` argument here:
```julia
        clf = models[1](
                builder = MLJFlux.Short(n_hidden = 5, dropout = 0.2),
                optimiser = Optimisers.Adam(0.01),
                batch_size = 8,
                epochs = 100,
                acceleration = CUDALibs(),
                optimiser_changes_trigger_retraining = true,
                embedding_dims = embedding_dims[3],
                rng=42
            )
```

It's needed to check that the output is the same with or wirhout the wrapper.

**Comment by ablaom** [comment]
Thanks @EssamWisam  for the diagnostics. That's very helpful. The good news is that a warning we added previously has correctly flagged the issue. 

The build kite tests include GPU tests, which is why you are not seeing the issue locally, I expect. Reproducibility using RNGs on a GPU is a can of worms, and we need to dodge that. The layers are initialised on the CPU and moved across. It's just the dropout that causes the problems. 

Can you please try this:

- get rid of the dropout (set to 0)
- in a prelminary step, define `stable_rng=StableRNG(123)`, and in your classifier set `rng=stable_rng` instead of seeding it.
- make a second classifier `clf2 = deepcopy(clf)`. Use `clf2` instead of `clf` in your second `machine` (line 204). (At present, when you fit using `clf`, it's `rng` hyperparameter is being mutated. By making a deep copy, we ensure `clf2` has an rng beginning in the same state. Probably, we should change `fit` to make a deep copies of the rng before use, to ensure the model is never mutated, but that's a discussion for another day.)
 
 Unrelated comment: It looks like `models[1]` was (at some time) meant to be `models[i]`. Is the more restrictive test intentional? Probably testing one of the models is sufficient, so I don't have a problem with this. I just noticed it.

**Comment by EssamWisam** [comment]
I did deep copying and no dropout as you said. When I use the `stable_rng` variable the equality fails so I left it as `42`.

**Comment by EssamWisam** [comment]
>  It looks like models[1] was (at some time) meant to be models[i].

Indeed, I can look into this later.

**Comment by ablaom** [comment]
Okay, sorry, we need something without a dropout layer at all. How about something like `builder = MLP(hidden=(10, 10))`?

**Comment by EssamWisam** [comment]
Done.

File: src/mlj_embedder_interface.jl [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+# 1. Interface Struct
+mutable struct EntityEmbedder{M <: MLJFluxModel} <: Unsupervised
+    model::M
+end;
+
+
+# 2. Constructor
+function EntityEmbedder(model;)
+    return EntityEmbedder(model)
```
Comment (ablaom) [inline src/mlj_embedder_interface.jl:11]: For consistency with other wrappers, please provide a keyword constructor `EntityEmbedder(; model=nothing)` as well. If the user doesn't specify (i.e., does `EntityEmbedder()`, so that `isnothing(model)`, then throw an ArgumentException("You must specify a suitable MLJFlux supervised model, as in \`EntityEmbedder(model=...)\`.  ") or similar.

File: src/mlj_embedder_interface.jl [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    package_name = "MLJTransforms",
+    package_uuid = "23777cdb-d90c-4eb0-a694-7c2b83d5c1d6",
+    package_url = "https://github.com/JuliaAI/MLJTransforms.jl",
+    is_pure_julia = true,
+)
+
+MMI.metadata_model(
+    EntityEmbedder,
+    input_scitype = Table,
+    output_scitype = Table,
```
Comment (ablaom) [inline src/mlj_embedder_interface.jl:42]:
```suggestion
```

Because this is a model wrapper, we need to forward a bunch of traits from the atomic model type to the wrapped model type. Something like this:

```julia
MMI.is_wrapper(::Type{<:EntityEmbedder}) =true
MMI..supports_training_losses(::Type{<:EntityEmbedder}) = true


# # TRAIT FORWARDING

for trait in [
    :input_scitype,
    :output_scitype,
    :target_scitype,
    ]

    quote
        MMI.$trait(::Type{<:EntityEmbdder{M}}) where M = MMI.$trait(M)
    end |> eval
end

# ## Iteration parameter

quote
    MMI.iteration_parameter(::Type{<:EntityEmbedder{M}}) where M =
        prepend(:model, MMI.iteration_parameter(M))
    end |> eval
end
```
You will need to duplicate the tool `prepend` from MLJModels; code is [here](https://github.com/JuliaAI/MLJModels.jl/blob/c6f536f26c9730f67d3cf083a84a71ae9a8ef90d/src/utilities.jl#L55).

File: src/mlj_embedder_interface.jl [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+Here:
+
+
+- `X` is any table of input features (eg, a `DataFrame`). Features to be transformed must
+   have element scitype `Multiclass` or `OrderedFactor`. Use `schema(X)` to 
+   check scitypes. 
+
+- `y` is the target, which can be any `AbstractVector` whose element
+  scitype is `Continuous` or `Count` for regression problems and
```
Comment (ablaom) [inline src/mlj_embedder_interface.jl:71]:
```suggestion
  scitype is `Continuous` or for regression problems and 
```

`Count` targets are not currently supported by the MLJFlux models you can wrap here.

Perhaps to prevent tragile wrapper docs, you can just say that `X` and `y` can have any type supported by the model  being wrapped?

File: src/mlj_embedder_interface.jl [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+- `y` is the target, which can be any `AbstractVector` whose element
+  scitype is `Continuous` or `Count` for regression problems and 
+  `Multiclass` or `OrderedFactor` for classification problems; check the scitype with `schema(y)`
+
+Train the machine using `fit!(mach)`.
+
+# Hyper-parameters
+
+- `model`: The underlying deep learning model to be used for entity embedding. So far this supports `NeuralNetworkClassifier`, `NeuralNetworkRegressor`, and `MultitargetNeuralNetworkRegressor`.
```
Comment (ablaom) [inline src/mlj_embedder_interface.jl:78]:
```suggestion
- `model`: The supervised MLJFlux neural network model to be used for entity embedding. 
  This must be one of these: `MLJFlux.NeuralNetworkClassifier`,
  `MLJFlux.NeuralNetworkRegressor`,`MLJFlux.MultitargetNeuralNetworkRegressor`.
```

File: src/mlj_embedder_interface.jl [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+# Operations
+
+- `transform(mach, Xnew)`: Transform the categorical features of `Xnew` into dense `Continuous` vectors using the trained `MLJFlux.EntityEmbedderLayer` layer present in the network.
+    Check relevant documentation [here](https://fluxml.ai/MLJFlux.jl/dev/) and in particular, the `embedding_dims` hyperparameter.
+
+
+# Examples
+
+```julia
+using MLJFlux
```
Comment (ablaom) [inline src/mlj_embedder_interface.jl:89]:
```suggestion
```

File: src/mlj_embedder_interface.jl [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    Column4 = repeat(Float32[1.0, 2.0, 3.0, 4.0, 5.0], Int(N / 5)),
+    Column5 = randn(Float32, N),
+    Column6 = categorical(
+        repeat(["group1", "group1", "group2", "group2", "group3"], Int(N / 5)),
+    ),
+)
+y = categorical([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])           # Classification
+
+# Initiate model
+NeuralNetworkClassifier = @load NeuralNetworkClassifier pkg=MLJFlux
```
Comment (ablaom) [inline src/mlj_embedder_interface.jl:108]:
```suggestion
EntityEmbedder = @load EntityEmbedder pkg=MLJFlux
NeuralNetworkClassifier = @load NeuralNetworkClassifier pkg=MLJFlux
```

File: src/mlj_embedder_interface.jl [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+MMI.target_in_fit(::Type{<:EntityEmbedder}) = true
+
+
+
+
+
+"""
+$(MMI.doc_header(EntityEmbedder))
+
```
Comment (ablaom) [inline src/mlj_embedder_interface.jl:54]: The "canned" header does not work here, for it includes "Do model = EntityEmbedder() to construct an instance with default hyper-parameters." but in case of a wrapper you must specify `model`. 

I also suggest adding a cautionary note somewhere, to the effect that for best results, the default `builder` hyperparameter for the atomic model may be ned to be explicitly set to an appropriate value, i.e the default may not be good enough.

**Review by ablaom — COMMENTED** [review]
Thanks for the great work. Please see what you can do with my suggestions.

File: src/mlj_embedder_interface.jl (hunk 1 above)
Comment (ablaom) [reply src/mlj_embedder_interface.jl:11]: For the logic, see for example your [`BalancedModel` code](https://github.com/JuliaAI/MLJBalancing.jl/blob/dev/src/balanced_model.jl).

File: src/mlj_embedder_interface.jl (hunk 2 above)
Comment (ablaom) [reply src/mlj_embedder_interface.jl:42]: And we need to defined the `training_losses` function:

```julia
MMI.training_losses(embedder::EntityEmbedder, report) =
    MMI.training_losses(embedder.model, report)
```

I think this works, because your `fit` is just outputting as `report` the atomic `report`. 

[Docs](https://juliaai.github.io/MLJModelInterface.jl/dev/reference/#MLJModelInterface.training_losses-Tuple%7BAny,%20Any%7D)

File: src/mlj_embedder_interface.jl (hunk 1 above)
Comment (EssamWisam) [reply src/mlj_embedder_interface.jl:13]: Done in baf10d9b8988952e95e9f41ddddb40fa5d8dbdb2

File: src/mlj_embedder_interface.jl (hunk 3 above)
Comment (EssamWisam) [reply src/mlj_embedder_interface.jl]: Done in 8ac77e4ed2cce4869dcdf98a8ebd57634cd47f06.

File: src/mlj_embedder_interface.jl (hunk 7 above)
Comment (EssamWisam) [reply src/mlj_embedder_interface.jl:56]: 0753e94f35ec73a72cb634e8c85452fdca05a6b5 tackles this.

File: src/mlj_embedder_interface.jl (hunk 2 above)
Comment (EssamWisam) [reply src/mlj_embedder_interface.jl]: Done in 34a5af3c8e9247f7871e22b76686c8b6d326f028

**Comment by EssamWisam** [comment]
@ablaom I took actions regarding all the points. Please check if it's ready now and thank you.

**Review by ablaom — APPROVED** [review]
This is great work, thanks. 

One final thing I noticed is that `NeuralNetworkBinaryClassifier` [is supported](https://github.com/FluxML/MLJFlux.jl/blob/34a5af3c8e9247f7871e22b76686c8b6d326f028/src/classifier.jl#L63) by the wrapper, but this is missing from the docs.

Otherwise good to go

File: src/mlj_embedder_interface.jl [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+   check scitypes. 
+
+- `y` is the target, which can be any `AbstractVector` supported by the model being wrapped.
+
+Train the machine using `fit!(mach)`.
+
+# Hyper-parameters
+
+- `model`: The supervised MLJFlux neural network model to be used for entity embedding. 
+  This must be one of these: `MLJFlux.NeuralNetworkClassifier`,
```
Comment (EssamWisam) [inline src/mlj_embedder_interface.jl:104]:
```suggestion
  This must be one of these: `MLJFlux.NeuralNetworkClassifier`, `NeuralNetworkBinaryClassifier`
```

File: src/mlj_embedder_interface.jl [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+   check scitypes. 
+
+- `y` is the target, which can be any `AbstractVector` supported by the model being wrapped.
+
+Train the machine using `fit!(mach)`.
+
+# Hyper-parameters
+
+- `model`: The supervised MLJFlux neural network model to be used for entity embedding. 
+  This must be one of these: `MLJFlux.NeuralNetworkClassifier`, `NeuralNetworkBinaryClassifier`
```
Comment (EssamWisam) [inline src/mlj_embedder_interface.jl:104]:
```suggestion
  This must be one of these: `MLJFlux.NeuralNetworkClassifier`, `NeuralNetworkBinaryClassifier`,
```

**Comment by EssamWisam** [comment]
> This is great work, thanks.
> 
> One final thing I noticed is that `NeuralNetworkBinaryClassifier` [is supported](https://github.com/FluxML/MLJFlux.jl/blob/34a5af3c8e9247f7871e22b76686c8b6d326f028/src/classifier.jl#L63) by the wrapper, but this is missing from the docs.
> 
> Otherwise good to go

I used to think that `NeuralNetworkBinaryClassifier` is redundant (and did not know its exposed in docs) because `NeuralNetworkClassifier` exists and Sigmoid function is a special case of Softmax when there are two classes. 

I do indeed wonder why both exist, now that I realize that the binary classifier is exposed in MLJFlux docs.

In any case, I added it in the docs here. Feel free to merge.

**Comment by ablaom** [comment]
> I used to think that NeuralNetworkBinaryClassifier is redundant (and did not know its exposed in docs) because NeuralNetworkClassifier exists and Sigmoid function is a special case of Softmax when there are two classes.

This classifier was added by @tiemvanderdeure . My understanding is that it provides some binary-specific optimisations, which was worth doing as  it is such a common use case.

**Comment by EssamWisam** [comment]
> > I used to think that NeuralNetworkBinaryClassifier is redundant (and did not know its exposed in docs) because NeuralNetworkClassifier exists and Sigmoid function is a special case of Softmax when there are two classes.
> 
> This classifier was added by @tiemvanderdeure . My understanding is that it provides some binary-specific optimisations, which was worth doing as it is such a common use case.

If a Sigmoid is used then I can imagine that the final matrix multiplication does becomes faster because the corresponding matrix now has one less column. That said, if it's true that they are mathematically equivalent, in the way I understand, then it may be better to automatically switch the `NeuralNetworkClassifier` to Sigmoid when there are two classes and only expose it.

Just sharing thoughts.

**Comment by tiemvanderdeure** [comment]
Yes, I added the binary classifier because it wasn't possible to use a sigmoid finalizer or binary cross-entropy loss with the existing classifier.

> That said, if it's true that they are mathematically equivalent, in the way I understand, then it may be better to automatically switch the `NeuralNetworkClassifier` to Sigmoid when there are two classes and only expose it.

I'm not against this at all. I wasn't aware that they are mathematically equivalent. I think I did some testing and got better results with the binary case (or it was just faster, I don't exactly remember).
